inskinesis/inskinesis.go

```diff
@@ -0,0 +1,393 @@
+}
+
+// stream represents a Kinesis stream and its properties.
+type stream struct {
+	region        string               // AWS region where the Kinesis stream is located.
+	name          string               // Name of the Kinesis stream.
+	partitioner   *PartitionerFunction // The partitioning function used to determine the partition key for records.
+	kinesisClient KinesisInterface     // AWS Kinesis client for interacting with the stream.
+
+	logBufferSize          int // Maximum size of the log buffer for records.
+	maxStreamBatchSize     int // Maximum size of each batch of records to be sent to the stream.
+	maxStreamBatchByteSize int // Maximum size (in bytes) of each batch of records.
+	maxGroup               int // Maximum number of concurrent groups for sending records.
+
+	mu               sync.Mutex         // Mutex to synchronize access to the stream.
+	wgLogChan        *sync.WaitGroup    // WaitGroup to manage goroutines.
+	wgBatchChan      *sync.WaitGroup    // WaitGroup to manage goroutines.
+	logChannel       chan interface{}   // Channel for receiving individual log records.
+	batchChannel     chan []interface{} // Channel for sending batches of log records.
+	stopChannel      chan bool          // Channel to signal the termination of the streaming process.
+	errChannel       chan error         // Channel for receiving errors.
+	stopBatchChannel chan bool          // Channel to signal the termination of batch streaming.
+	logBuffer        []interface{}      // Buffer for accumulating log records before batching.
+
+	failedCount int // Counter for the number of failed record submissions.
+	totalCount  int // Counter for the total number of records sent to the stream.
+}
+
+type Config struct {
+	Region                 string
+	StreamName             string
+	Partitioner            *PartitionerFunction
+	MaxStreamBatchSize     int
+	MaxStreamBatchByteSize int
+	MaxBatchSize           int
+	MaxGroup               int
+}
+
+// NewKinesis creates a new Kinesis stream.
+func NewKinesis(config Config) (StreamInterface, error) {
+	if config.Region == "" {
+		return nil, errors.New("region is required")
+	}
+
+	if config.StreamName == "" {
+		return nil, errors.New("stream name is required")
+	}
+	awsConfig := aws.Config{Region: aws.String(config.Region)}
+	awsSession, err := session.NewSession(&awsConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	kinesisClient := kinesis.New(awsSession)
+
+	s := &stream{
+		region:        config.Region,
+		name:          config.StreamName,
+		partitioner:   config.Partitioner,
+		kinesisClient: kinesisClient,
+
+		logBufferSize:          config.MaxBatchSize,
+		maxStreamBatchSize:     config.MaxStreamBatchSize,
+		maxStreamBatchByteSize: config.MaxStreamBatchByteSize,
+		maxGroup:               config.MaxGroup,
+
+		wgLogChan:        &sync.WaitGroup{},
+		wgBatchChan:      &sync.WaitGroup{},
+		logChannel:       make(chan interface{}, 1000),
+		batchChannel:     make(chan []interface{}, 100),
```

Comment:
TODO: Make these configurable in the future

Reply:
buna sonra girelim gerekirse